File: app/src/main/java/com/example/flicks/DetailActivity.java
package com.example.flicks;

import android.annotation.SuppressLint;
import android.os.Bundle;
import android.widget.ImageView;
import android.widget.RatingBar;
import android.widget.TextView;

import com.bumptech.glide.Glide;
import com.bumptech.glide.request.RequestOptions;
import com.example.flicks.api.ApiModule;
import com.example.flicks.api.ApiService;
import com.example.flicks.entity.Videos;
import com.example.flicks.mapper.DataMapper;
import com.example.flicks.model.Movie;
import com.example.flicks.model.Trailer;
import com.google.android.youtube.player.YouTubeBaseActivity;
import com.google.android.youtube.player.YouTubeInitializationResult;
import com.google.android.youtube.player.YouTubePlayer;
import com.google.android.youtube.player.YouTubePlayerView;

import java.util.List;

import jp.wasabeef.glide.transformations.RoundedCornersTransformation;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

import static com.bumptech.glide.request.RequestOptions.bitmapTransform;

public class DetailActivity extends YouTubeBaseActivity {

    ImageView img_poster_detail;
    TextView tv_title;
    TextView tv_release_date;
    TextView tv_detail_release_date;
    RatingBar rating_bar;
    TextView tv_overview_detail;
    YouTubePlayerView youtube_player;
    TextView tv_rating;

    protected List<Trailer> trailersList;
    protected DataMapper mapper;
    protected Movie movie;

    @Override
    protected void onCreate( Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.detail_activity);

        img_poster_detail = findViewById(R.id.img_poster_detail);
        tv_title = findViewById(R.id.tv_title);
        tv_release_date = findViewById(R.id.tv_release_date);
        tv_detail_release_date = findViewById(R.id.tv_detail_release_date);
        rating_bar = findViewById(R.id.rating_bar);
        tv_overview_detail = findViewById(R.id.tv_overview_detail);
        youtube_player = findViewById(R.id.youtube_player);
        tv_rating = findViewById(R.id.tv_rating);

        // lấy dữ liệu từ bên màn hình cũ
        Bundle bundle = getIntent().getExtras();
        movie = (Movie) (bundle != null ? bundle.getParcelable("movie") : null);

        setupData();
        getTrailer();
    }

    private void getTrailer() {

        mapper = new DataMapper();

        // lấy link video trailer youtube
        ApiModule.getInstance().getVideo(movie.getId(), ApiService.API_KEY)
                .enqueue(new Callback<Videos>() {
                    @Override
                    //gọi thành công
                    public void onResponse(Call<Videos> call, Response<Videos> response) {
                        if(response.body() != null){
                            trailersList = mapper.transform(response.body());
                            String url = trailersList.get(0).getKey();
                            youtubePlayer(url);

                        }
                    }
                    @Override
                    //gọi thất bại
                    public void onFailure(Call<Videos> call, Throwable t) {
                    }
                });

    }

    private void youtubePlayer(final String key) {
        //setup link video để chạy youtube
        youtube_player.initialize(ApiService.API_KEY_YOUTUBE,
                new YouTubePlayer.OnInitializedListener() {
                    @Override
                    //set link thành công
                    public void onInitializationSuccess(YouTubePlayer.Provider provider, YouTubePlayer youTubePlayer, boolean b) {
                        // nếu rating phim > 7.5 thì play video luôn không thì chỉ load
                        if (movie.getVoteAverage() >= ApiService.AVER_RATING) {
                            youTubePlayer.loadVideo(key); // chạy video
                            youTubePlayer.setShowFullscreenButton(false);
                        } else youTubePlayer.cueVideo(key); //load video
                    }

                    @Override
                    //load video thất bại
                    public void onInitializationFailure(YouTubePlayer.Provider provider, YouTubeInitializationResult youTubeInitializationResult) {

                    }
                });
    }

    @SuppressLint("SetTextI18n")
    private void setupData() {

        //set dữ liệu phim vào từng view
        tv_title.setText(movie.getTitle());
        rating_bar.setRating(movie.getVoteAverage().floatValue());
        rating_bar.setIsIndicator(true);
        tv_detail_release_date.setText(movie.getReleaseDate());
        tv_overview_detail.setText(movie.getOverview());
        tv_rating.setText(movie.getVoteAverage().toString());

        // xài thư viện Glide để load hình
        Glide.with(DetailActivity.this).load(movie.getPosterPath()).apply(new RequestOptions()
                .fitCenter())
                .apply(bitmapTransform(new RoundedCornersTransformation(10, 0, RoundedCornersTransformation.CornerType.ALL)))
                .into(img_poster_detail);

    }
}

File: app/src/main/java/com/example/flicks/DetailActivity.kt
package com.example.flicks

import android.annotation.SuppressLint
import android.os.Bundle
import android.os.Parcelable
import android.widget.ImageView
import android.widget.RatingBar
import android.widget.TextView
import com.bumptech.glide.Glide
import com.bumptech.glide.request.RequestOptions
import com.example.flicks.api.ApiModule
import com.example.flicks.api.ApiService
import com.example.flicks.entity.Videos
import com.example.flicks.model.Movie
import com.example.flicks.model.Trailer
import com.google.android.youtube.player.YouTubeBaseActivity
import com.google.android.youtube.player.YouTubeInitializationResult
import com.google.android.youtube.player.YouTubePlayer
import com.google.android.youtube.player.YouTubePlayerView
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response

class DetailActivity : YouTubeBaseActivity() {
    var img_poster_detail: ImageView? = null
    var tv_title: TextView? = null
    var tv_release_date: TextView? = null
    var tv_detail_release_date: TextView? = null
    var rating_bar: RatingBar? = null
    var tv_overview_detail: TextView? = null
    var youtube_player: YouTubePlayerView? = null
    var tv_rating: TextView? = null
    private var trailersList: List<Trailer>? = null
    private var movie: Movie? = null
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.detail_activity)
        img_poster_detail = findViewById(R.id.img_poster_detail)
        tv_title = findViewById(R.id.tv_title)
        tv_release_date = findViewById(R.id.tv_release_date)
        tv_detail_release_date = findViewById(R.id.tv_detail_release_date)
        rating_bar = findViewById(R.id.rating_bar)
        tv_overview_detail = findViewById(R.id.tv_overview_detail)
        youtube_player = findViewById(R.id.youtube_player)
        tv_rating = findViewById(R.id.tv_rating)

        // lấy dữ liệu từ bên màn hình cũ
        val bundle = intent.extras
        movie = bundle?.getParcelable<Parcelable>("movie") as Movie?
        setupData()
        getTrailer()
    }

    // lấy link video trailer youtube
    private fun getTrailer() {
        // lấy link video trailer youtube
        ApiModule.instance?.getVideo(movie!!.id, ApiService.API_KEY)
                ?.enqueue(object : Callback<Videos?> {
                    //gọi thành công
                    override fun onResponse(call: Call<Videos?>, response: Response<Videos?>) {
                        if (response.body() != null) {
                            val url = response.body()?.results?.get(0)?.key
                            youtubePlayer(url)
                        }
                    }

                    override fun onFailure(call: Call<Videos?>, t: Throwable) {

                    }
                })
    }

    private fun youtubePlayer(key: String?) {
        //setup link video để chạy youtube
        youtube_player!!.initialize(ApiService.API_KEY_YOUTUBE,
                object : YouTubePlayer.OnInitializedListener {
                    //set link thành công
                    override fun onInitializationSuccess(provider: YouTubePlayer.Provider, youTubePlayer: YouTubePlayer, b: Boolean) {
                        // nếu rating phim > 7.5 thì play video luôn không thì chỉ load
                        if (movie!!.voteAverage >= ApiService.AVER_RATING) {
                            youTubePlayer.loadVideo(key) // chạy video
                            youTubePlayer.setShowFullscreenButton(false)
                        } else youTubePlayer.cueVideo(key) //load video
                    }

                    //load video thất bại
                    override fun onInitializationFailure(provider: YouTubePlayer.Provider, youTubeInitializationResult: YouTubeInitializationResult) {}
                })
    }

    @SuppressLint("SetTextI18n")
    private fun setupData() {

        //set dữ liệu phim vào từng view
        tv_title!!.text = movie!!.title
        rating_bar!!.rating = movie!!.voteAverage.toFloat()
        rating_bar!!.setIsIndicator(true)
        tv_detail_release_date!!.text = movie!!.releaseDate
        tv_overview_detail!!.text = movie!!.overview
        tv_rating!!.text = movie!!.voteAverage.toString()

        // xài thư viện Glide để load hình
        Glide.with(this@DetailActivity).load(movie!!.posterUrl()).apply(RequestOptions()
                .fitCenter())
                .into(img_poster_detail!!)
    }
}